### PR TITLE
Added tune method to API

### PR DESCRIPTION
**Pull request by aPovidlo**
Solution to Issue #1018

### CODE REVIEW

File: fedot/api/main.py [hunk 1]
```diff
@@ -200,6 +203,60 @@ def fit(self,
 
         return self.current_pipeline
 
+    def tune(self,
+             metric_name: str = None,
+             iterations: int = DEFAULT_TUNING_ITERATIONS_NUMBER,
+             timeout: Optional[float] = 5,
+             cv_folds: int = None,
+             n_jobs: int = 1,
+             show_progress: bool = False) -> Pipeline:
```
Comment (YamLyubov) [inline fedot/api/main.py:212]: Тут нужно добавить еще ``validation_blocks`` иначе тюнер будет расчитывать метрику для временных рядов не так, как до этого композер

File: test/unit/api/test_main_api.py [hunk 2]
```diff
@@ -164,19 +165,41 @@ def test_api_predict_correct(task_type, predefined_model, metric_name):
...
+])
+def test_api_tune_correct(task_type, metric_name, pred_model):
+    task_type = task_type.value
 
-    model.fit(features=train_data, predefined_model='auto')
-    ts_forecast = model.predict(features=test_data)
-    _ = model.get_metrics(target=test_data.target, metric_names='rmse')
+    if task_type is 'ts_forecasting':
+        forecast_length = 5
+        train_data, test_data, _ = get_dataset(task_type, validation_blocks=1)
```
Comment (YamLyubov) [inline test/unit/api/test_main_api.py:178]: Если здесь ```validation_blocks=1``` тогда в тюнере они тоже должны быть равны единице. Сейчас в тюнере ``validation_blocks=None``, что не тоже самое

File: test/unit/api/test_main_api.py [hunk 3]
```diff
@@ -164,19 +165,41 @@ def test_api_predict_correct(task_type, predefined_model, metric_name):
...
+    else:
+        train_data, test_data, _ = get_dataset(task_type)
+        model = Fedot(problem=task_type, timeout=0.1)
 
-    assert len(ts_forecast) == forecast_length
+    base_pipeline = deepcopy(model.fit(features=train_data, predefined_model=pred_model))
+    pred_before = model.predict(features=test_data)
+    metric_before = model.get_metrics()
+
+    tuned_pipeline = deepcopy(model.tune(timeout=0.5))
```
Comment (YamLyubov) [inline test/unit/api/test_main_api.py:191]: Можно уменьшить таймаут до 0.1

**Review by YamLyubov — CHANGES_REQUESTED** [review]

File: fedot/api/main.py [hunk 4]
```diff
@@ -200,6 +203,48 @@ def fit(self,
...
+        """
+        if self.current_pipeline is None:
+            raise ValueError(NOT_FITTED_ERR_MSG)
+
+        if metric_name is not None:
+            metric = self.metrics.get_metrics_mapping(metric_name)
+        elif self.params.metric_to_tuning is not None:
+            metric = self.params.metric_to_tuning
+        elif self.params.metric_to_compose is not None:
+            metric = self.params.metric_to_compose
```
[the remark this answers is not in the document]
Comment (aPovidlo) [reply fedot/api/main.py:235]: @YamLyubov @nicl-nno А их как-то может починить или убрать забыли когда-то?

File: fedot/api/main.py [hunk 5]
```diff
@@ -200,6 +203,60 @@ def fit(self,
...
+            .with_tuner(PipelineTuner) \
+            .with_iterations(iterations) \
+            .with_cv_folds(cv_folds) \
+            .with_n_jobs(n_jobs) \
+            .with_metric(metric) \
+
+        if timeout in [-1, None]:
+            pipeline_tuner = pipeline_tuner.with_timeout(timeout=None)
+        else:
+            pipeline_tuner = pipeline_tuner.with_timeout(timedelta(seconds=60 * timeout))
```
[the remark this answers is not in the document]
Comment (aPovidlo) [reply fedot/api/main.py:250]: Да, это я почему-то подумал, что timedelta() не будет воспринимать float. Проверил и он все-таки будет, поэтому исправлю.

File: fedot/api/main.py [hunk 6]
```diff
@@ -200,6 +203,60 @@ def fit(self,
...
+
+        pipeline_tuner = TunerBuilder(self.train_data.task) \
+            .with_tuner(PipelineTuner) \
+            .with_iterations(iterations) \
+            .with_cv_folds(cv_folds) \
+            .with_n_jobs(n_jobs) \
+            .with_metric(metric) \
+
+        if timeout in [-1, None]:
+            pipeline_tuner = pipeline_tuner.with_timeout(timeout=None)
```
[the remark this answers is not in the document]
Comment (aPovidlo) [reply fedot/api/main.py:248]: Это скорее было сделанно, потому что если не указывать время, то будет дефолтное значение при инициализации. Думаю, что может быть по умолчанию бесконечность ставить, а не в методе изменять?

File: fedot/api/main.py [hunk 7]
```diff
@@ -200,6 +203,60 @@ def fit(self,
...
+        else:
+            pipeline_tuner = pipeline_tuner.with_timeout(timedelta(seconds=60 * timeout))
+
+        pipeline_tuner = pipeline_tuner.build(self.train_data)
+
+        self.current_pipeline = pipeline_tuner.tune(self.current_pipeline, show_progress)
+        # Tuner returns a not fitted pipeline, and it is required to fit on train dataset
+        self.current_pipeline.fit(self.train_data)
```
[the remark this answers is not in the document]
Comment (aPovidlo) [reply fedot/api/main.py:256]: А что значит > новые для тюнинга?

File: fedot/api/main.py (hunk 1 above)
Comment (aPovidlo) [reply fedot/api/main.py:212]: Спасибо, исправлю это

File: test/unit/api/test_main_api.py (hunk 3 above)
Comment (aPovidlo) [reply test/unit/api/test_main_api.py:191]: Хорошо

File: fedot/api/main.py (hunk 1 above)
Comment (aPovidlo) [reply fedot/api/main.py:212]: @YamLyubov А он же обязательно должен быть одинаковым с тем, который был указан в композере, или не всегда?

File: fedot/api/main.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (nicl-nno) [reply fedot/api/main.py:235]: Видимо забыли убрать - можно в рамках этого PR это и сделать (надо только проверит).

File: fedot/api/main.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (YamLyubov) [reply fedot/api/main.py:256]: Ну то есть мы получили модель в рамках Fedot.fit на одних данных, а потом решили подобрать гиперпараметры через тюнинг на других данных.

File: fedot/api/main.py (hunk 1 above)
Comment (YamLyubov) [reply fedot/api/main.py:212]: От количества ```validation_blocks`` зависит то, как мы будем рассчитывать метрику.

Comment (YamLyubov) [reply fedot/api/main.py:212]: Также get_metrics() тоже использует ```validation_blocks``` тоже, что и композер.

File: fedot/api/main.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (YamLyubov) [reply fedot/api/main.py:248]: Тогда в TunerBuilder надо еще поменять ```self.iterations=DEFAULT_TUNING_ITERATIONS_NUMBER``` и ```self.timeout=None```. В hyperopt нельзя прям бесконечность поставить

File: test/unit/api/test_main_api.py (hunk 2 above)
Comment (aPovidlo) [reply test/unit/api/test_main_api.py:178]: Исправил

File: fedot/api/main.py (hunk 1 above)
Comment (aPovidlo) [reply fedot/api/main.py:212]: Думаю, что у меня получилось это исправить, если обернуть метод в `api_composer` чтобы использовать информацию из `composer_requirements`

File: fedot/api/main.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (aPovidlo) [reply fedot/api/main.py:248]: Исправил все это в `tuner_builder.py`

File: fedot/api/main.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (aPovidlo) [reply fedot/api/main.py:235]: Убрал

File: fedot/api/api_utils/api_composer.py [hunk 8]
```diff
@@ -288,6 +291,27 @@ def compose_pipeline(self, task: Task,
         best_pipeline = best_pipelines[0] if isinstance(best_pipelines, Sequence) else best_pipelines
         return best_pipeline, best_pipeline_candidates, gp_composer
 
+    def tune_pipeline(self, pipeline: Pipeline,
```
Comment (YamLyubov) [inline fedot/api/api_utils/api_composer.py:294]: Мне кажется, этот метод дублирует ```tune_final_pipeline```, было бы лучше этот метод все-таки в api вынести. Можно же просто тогда доставать ```composer_params = self.api_composer.composer_requirements```

File: fedot/api/main.py [hunk 9]
```diff
@@ -200,6 +200,63 @@ def fit(self,
...
+        params = {
+            "metric_name": metric,
+            "iterations": iterations,
+            "timeout": timeout,
+            "cv_folds": cv_folds,
+            "n_jobs": n_jobs,
+            "show_progress": show_progress
+        }
+
+        self.current_pipeline = ApiComposer.tune_pipeline(
```
Comment (YamLyubov) [inline fedot/api/main.py:247]: Тут скорее всего имелось в виду ```self.api_composer```

**Review by YamLyubov — CHANGES_REQUESTED** [review]

File: fedot/api/main.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nicl-nno) [reply fedot/api/main.py:256]: Если реализовал - пусть будет, в теории может пригодится. Главное что их передача необязательна.

File: fedot/api/api_utils/api_composer.py [hunk 10]
```diff
@@ -337,7 +349,7 @@ def _divide_parameters(common_dict: dict) -> List[dict]:
                                 optimizer_external_params=None, collect_intermediate_metric=False,
                                 max_pipeline_fit_time=None, initial_assumption=None, preset='auto',
                                 use_pipelines_cache=True, use_preprocessing_cache=True, cache_folder=None,
-                                keep_history=True, history_dir=None,)
+                                keep_history=True, history_dir=None, )
```
Comment (nicl-nno) [inline fedot/api/api_utils/api_composer.py:352]: Лишняя запятая?

File: fedot/api/main.py [hunk 11]
```diff
@@ -200,6 +205,58 @@ def fit(self,
...
+            raise ValueError(NOT_FITTED_ERR_MSG)
+
+        if input_data is None:
+            input_data = self.train_data
+
+        if metric_name is not None:
+            metric = self.metrics.get_metrics_mapping(metric_name)
+        else:
+            # Take first problems metrics
+            metric = self.metrics.get_metrics_mapping(self.metrics.get_problem_metrics()[0])
```
Comment (nicl-nno) [inline fedot/api/main.py:241]: Кажется, можно часть metric = self.metrics.get_metrics_mapping вынести за if.

File: fedot/api/main.py [hunk 12]
```diff
@@ -200,6 +205,58 @@ def fit(self,
...
+        """Method for hyperparameters tuning of current pipeline
+
+        Args:
+            input_data: data for tuning pipeline
+            metric_name: name of metric for quality tuning
+            iterations: numbers of tuning iterations
+            timeout: time for tuning (in minutes). If ``None`` or ``-1`` means tuning until max iteration reach
+            cv_folds: number of folds on data for cross-validation.
+            n_jobs: num of ``n_jobs`` for parallelization (``-1`` for use all cpu's)
+            show_progress: shows progress of tuning if true
```
[the remark this answers is not in the document]
Comment (aPovidlo) [reply fedot/api/main.py:225]: В [тесте](https://github.com/aimclub/FEDOT/blob/9ea1b8c0fb7d38b7d677b203b1c6468e7a5805cd/test/unit/api/test_main_api.py#L173) можно увидеть, что передается только время на тюнинг. Из `fit` нам только необходимо знать параметр для метода `.with_requirements(...)`.

[the remark this answers is not in the document]
Comment (aPovidlo) [reply fedot/api/main.py:225]: Заметил, что `self.api_composer.tuner_requirements` и `cv_folds` и `n_jobs` перекрывают друг друга. Поэтому в следующем коммите исправлю этот момент

File: fedot/api/main.py (hunk 11 above)
Comment (aPovidlo) [reply fedot/api/main.py:241]: И вправду, можно вынести

File: fedot/api/main.py [hunk 13]
```diff
@@ -200,6 +205,61 @@ def fit(self,
...
+
+        Returns:
+            Pipeline object
+
+        """
+        if self.current_pipeline is None:
+            raise ValueError(NOT_FITTED_ERR_MSG)
+
+        if input_data is None:
+            input_data = self.train_data
```
Comment (YamLyubov) [inline fedot/api/main.py:235]: Может в этой и аналогичных строчках лучше переформатировать ```input_data = input_data or self.train_data```

File: fedot/api/main.py [hunk 14]
```diff
@@ -200,6 +205,61 @@ def fit(self,
...
+
+        if metric_name is None:
+            metric_name = self.metrics.get_problem_metrics()[0]
+
+        if cv_folds is None:
+            cv_folds = self.api_composer.tuner_requirements.cv_folds
+
+        if n_jobs is None:
+            n_jobs = self.api_composer.tuner_requirements.n_jobs
```
Comment (YamLyubov) [inline fedot/api/main.py:244]: Эти строчки дублируют логику TunerBuilder. Если ты делаешь такую замену, то уже нет смысла в билдер и requirements, и cv_folds и n_jobs передавать. 

Было бы удобнее в самих ComposerRequirements заменить необходимые параметры и только их передавать

File: test/unit/api/test_main_api.py [hunk 15]
```diff
@@ -164,19 +165,41 @@ def test_api_predict_correct(task_type, predefined_model, metric_name):
...
-    train_data, test_data, _ = get_dataset(task_type, validation_blocks=1)
-    model = Fedot(problem='ts_forecasting', **default_params,
-                  task_params=TsForecastingParams(forecast_length=forecast_length),
-                  validation_blocks=1)
+@pytest.mark.parametrize('task_type, metric_name, pred_model', [
+    (TaskTypesEnum.classification, 'f1', 'dt'),
+    (TaskTypesEnum.regression, 'rmse', 'dtreg'),
+    (TaskTypesEnum.ts_forecasting, 'rmse', 'auto')
+])
+def test_api_tune_correct(task_type, metric_name, pred_model):
```
Comment (YamLyubov) [inline test/unit/api/test_main_api.py:173]: У меня этот тест не проходит. Видимо модель переобучается на трейне.
image

**Review by YamLyubov — CHANGES_REQUESTED** [review]

File: fedot/api/main.py (hunk 13 above)
Comment (aPovidlo) [reply fedot/api/main.py:235]: Да, так будет лучше

File: fedot/api/main.py (hunk 14 above)
Comment (aPovidlo) [reply fedot/api/main.py:244]: Соглашусь

**Review by nicl-nno — APPROVED** [review]

**Review by YamLyubov — APPROVED** [review]
